### PR TITLE
New Portfile for "HackTV" project.

### PR DESCRIPTION
**Pull request by kesterlester**
#### Description

This patch provides a Portfile for an applciation called "hacktv" which can be used to make sofware defined radios (SDRs) act as television transmitters.

###### Type(s)
- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 12.6 21G115 arm64
Xcode 13.4.1 13F100

### CODE REVIEW

**Comment by catap** [comment]
@kesterlester thanks for contribution. I'd like to suggest:
 - squash all commits into one (you may use `git rebase -i`);
 - use the commit message like: `hacktv: submission (version 0.0.4)`

Also, I think that the right group is `multimedia` not `common` ;)

BTW any reason why you had used your fork and not upstream? Upstream seems alive (last merge 8 days ago).

**Comment by kesterlester** [comment]
Dear @catap ,

Some replies inline below.

> @kesterlester thanks for contribution. I'd like to suggest:
> 
> * squash all commits into one (you may use `git rebase -i`);
> * use the commit message like: `hacktv: submission (version 0.0.4)`

ok

> Also, I think that the right group is `multimedia` not `common` ;)

I selected 'comms' thinking this stood for 'communications'. (My memory may be faulty on the next comment, but I seem to also recall seeing some 'software defined radio' things in 'comms' which, rightly or wrongly reinforced my belief that 'comms' meant 'communications'.)

But if you can confirm that 'comms' in fact means 'commons' I would happily move the group as per your suggestion.

> BTW any reason why you had used your fork and not upstream? Upstream seems alive (last merge 8 days ago).

This would be ignorance on my part. I have not submitted a port file to Macports before, and was trying to follow what I understood to be the instructions on how to do this on the macports website
(specifically 7.3.1 of https://guide.macports.org/#project.contributing). The latter guidance doesn't explicitly mention from where a pull request should come (or at least if it does I can't find it).

If you are able to point me via web link to some instructions which explain how I should be using upstream, I will be happy to read them and follow them.

**Comment by kesterlester** [comment]
Correction to my last comment:  looking back into the repository, I can see that the files I saw in the comms directory that I mentioned being connected with software define radios were in fact connected with a different form of communication: "telepathy". When I saw the keys I thought oh yes telepathy that must be referring to a communication package albeit not one related to mine to mine for transfer :-)

Actually, software define radio things seem to be in the group called "science". So it may well be that the best time for this is science rather than multimedia. But I'm happy to take advice either way

**Comment by kesterlester** [comment]
Ah - I've just figured out what  @catap  you meant by your comment about my use of my rep rather than upstream.

I thought you were referring to my use of my fork of the macports-ports repository, but I've just realised that you are (probably) talking instead about my use of my fork of the original hacktv repository.

If that is what you were asking about, then the answer is that there were no **super-strong** reasons, but there were two weaker reasons:

1. Macports guidance seemed to indicate that it was preferred for github checkouts to be checkouts of releases rather than of random tags. The upstream hacktv maintainer appears not to make releases, but I thought I could "easily" do these on my own fork.
2.  I thought I would find it easier to directly patch my hacktv fork rather than apply patches (to the upstream) within the macports machinery. The former I felt confident to do, the latter I suspected would take me longer to understand how to do.

**Comment by kesterlester** [comment]
OK:

- I moved the Portfile from the comms group to the multimedia group, and,
- I also squashed the commits down to a single commit, but for reasos I won't explain (unless someone needs to hear them) the squashed commit has ended up in a new branch of my fork here:

https://github.com/kesterlester/macports-ports/tree/kesterlester-patch-2

So in summary, if the macports people want to have this "hacktv" port, they can either:

1. accept the current unsquashed kesterlester-patch-1 merge request, 

or

2. if they prefer a squashed commit they can ask me to make a new merge request from the kesterlester-patch-2 branch I have sitting currently in my fork at

https://github.com/kesterlester/macports-ports/tree/kesterlester-patch-2

I could of course issue the merge request for the latter now, but I didn't want to do so without being asked, as then we will end up with two in the system where perhaps only one is required.

**Comment by catap** [comment]
@kesterlester as far as I see squashing is MacPorts policy or guideline: https://guide.macports.org/#project.github

I've compared differences between your fork and upstream: https://github.com/fsphil/hacktv/compare/master...kesterlester:hacktv:master and I think that you may switch to upstream instead of fork, with adoption your Makefile's patch inside `destroot` block. 

I also support numberos of ports without any release on github, for example [IMAPdedup](https://github.com/macports/macports-ports/blob/master/mail/imapdedup/Portfile); I suggest to use the same idea here.

Also, if you make a force push to your branch, you may reuse this PR ;)

File: science/hacktv/Portfile
Comment (catap) [inline science/hacktv/Portfile:13]: Can you follow the maintainers keyword format?

https://guide.macports.org/#reference.keywords.maintainers

File: science/hacktv/Portfile
Comment (reneeotten) [inline science/hacktv/Portfile:7]: I agree with @catap this should come from `fsphil hacktv` not your fork. You can remove the `github.tarball_from` and just use a commit hash and specify a (likely) date-based version in the Portfile.

File: science/hacktv/Portfile
Comment (reneeotten) [inline science/hacktv/Portfile:11]: this is the default since MacPorts v2.7 so the `platforms` line should be removed

File: science/hacktv/Portfile
Comment (reneeotten) [inline science/hacktv/Portfile:16]: the homepage will be set automatically by the `github` PG so this line should be removed

File: science/hacktv/Portfile
Comment (reneeotten) [inline science/hacktv/Portfile:26]: this is problematic as it will likely not be ["UsingTheRightCompiler"](https://trac.macports.org/wiki/UsingTheRightCompiler). If there is not a configure phase you'll have to take care of that yourself, often the `makefile` PortGroup can help.

**Review by reneeotten — CHANGES_REQUESTED** [review]
also please squash the commits and amend the commit message so that it follows our guidelines (i.e., `<portname>: short description`).

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: Done in 48b46f0e0a

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: Fixed in 32fca62

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: Fixed in 068e26a

File: science/hacktv/Portfile
Comment (reneeotten) [inline science/hacktv/Portfile:7]: no, this should use upstream repository not your fork of it

File: science/hacktv/Portfile
Comment (reneeotten) [reply science/hacktv/Portfile]: not completely, it should include both an e-mail address and a GitHub handle; additionally, we strongly prefer to also add `openmaintainer` such that minor changes to the Portfile can be made without your explicit approval

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile:7]: yes, I know, I'm not there yet. I'm working in a private rep. I'll push here when I 'm done.  I'm fixing things peacemeal.  Don't review again until I say it's ready! :)

Comment (reneeotten) [reply science/hacktv/Portfile:7]: you can always request upstream to at least tag a version or ideally even start doing releases

Comment (reneeotten) [reply science/hacktv/Portfile:7]: > yes, I know, I'm not there yet. I'm working in a private rep. I'll push here when I 'm done. I'm fixing things peacemeal. Don't review again until I say it's ready! :)

okay, I'll mark this as draft then. Once you're done with it you can set it to "ready for review"

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: Hopefully ce99322956b9464d5eb383b6cb71b6cc58ee35a5 now solves this part. Still working on other parts.

File: science/hacktv/Portfile
Comment (reneeotten) [reply science/hacktv/Portfile:7]: > > yes, I know, I'm not there yet. I'm working in a private rep. 

FYI you're working in your fork but as soon as you push there this PR will get updated as well

Comment (kesterlester) [reply science/hacktv/Portfile:7]: Well, I'm working in two places, a private rep AND my fork.  But that aside, I see you marked this rep as "work in progress".  I didn't know that possibility existed. I'm hoping that with that enablement (thank you!) then you and others should not be bothered anymore .... even if I do push to my fork .... until such time as I mark things no longer "in progress".

Is that correct?

I don't want to cause you (or others) unnecessary noise.

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: This is now done (in  199b2b38e0f7d34dd626218fb856dd3a62a86c07  ) but am still working on other things.

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: thank you

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile:32]: In 2a74cb1feeb I ask the portfile to export ${configure.cc}  at build time to an environment variable picked up by the make file.

I have checked the exported variable is being used.

I hope this means that this is now fixed.

File: science/hacktv/files/destdir-fix.diff
Comment (catap) [inline science/hacktv/files/destdir-fix.diff:6]: why not use `$(CC)` here?

**Comment by kesterlester** [comment]
> thanks for the submittion @kesterlester! A few comments below; while making the changes also please squash the commits and amend the commit message so that it follows our guidelines (i.e., `<portname>: short description`).

Should all now be done.

File: science/hacktv/Portfile
Comment (catap) [inline science/hacktv/Portfile:30]: You may avoid this line by adding:
```tcl
PortGroup        makefile 1.0
```

File: science/hacktv/files/destdir-fix.diff
Comment (catap) [reply science/hacktv/files/destdir-fix.diff:6]: Thus, you may keep `$(CROSS_HOST)gcc` as default value:
```Makefile
CC ?= $(CROSS_HOST)gcc
```

Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:6]: I wanted to minimally alter the upstream person's makefile. I suppose I could have deleted line 5 altogether and it may have hand the same effect as what I did. But I just did the same thing this other way.  

There is an openmaintainer added, so you are welcome to change it if you wish.

File: science/hacktv/files/destdir-fix.diff
Comment (catap) [inline science/hacktv/files/destdir-fix.diff:9]: By using `?=` here instead of `:=` you may enjoy MacPorts cflags which probably makes this port universal.

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile:30]: Am guessing that would not add my lines 

```
build.args-append       MACPORTS_CC=${configure.cc} \
                        CXX=${configure.cxx} \
                        CPP=${configure.cpp}
```

but would instead add something like

```
build.args-append       CC=${configure.cc} \
                        CXX=${configure.cxx} \
                        CPP=${configure.cpp}
```
meaning that I'd then need to change the makefile diff, which is a more complex thing to maintain than this.

Look, I don't know. 

This is not really my thing. I should be writing a lecture for some undergrads I have to lecture to tomorrow morning. I'm not really that into code golf.  Am very happy to fix real weaknesses, security holes, supremely non-portable statements.  But when it's just about aesthetics I'm running out of motivation to carry on.

File: science/hacktv/files/destdir-fix.diff
Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:9]: Are you able to do this? I'm not "fsphil" (the author of that makefile) so I don't really want to make changes to his code that I don't fully understand. If you feel qualified to make those changes, I suppose you can.

What you say ("which probably makes this port universal") certainly sounds good!   It's just a bit beyond my personal competence level to judge ....

Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:9]: OK - I've just looked up on google what the difference between := and ?= is.  Based on that I can see that what you say sounds plausible ... ish.

File: science/hacktv/files/destdir-fix.diff
Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:6]: Ah - my last post must have crossed your one explaining:

`CC ?= $(CROSS_HOST)gcc`

which I have only just now seen.  Let me think about this.

Comment (catap) [reply science/hacktv/files/destdir-fix.diff:6]: @kesterlester I usually backport to upstream all my patches which I did to my ports here. Near all of them usually merged :)

I'd like to suggest to do the same way here: adjust Makefile as you needed and backport it as PR.

If one day such PR is merged, you simple remove patch :)

File: science/hacktv/Portfile
Comment (catap) [reply science/hacktv/Portfile:30]: @kesterlester if you move forward with my suggestion to patch Makefile when you may remove it ;)

Comment (kesterlester) [reply science/hacktv/Portfile:30]: OK - I saw that now.  I think I understand what's going on.

But one thing I don't understand is that this merge request has suddenly gone "closed".

I haven't yet managed figure out whether that means it was rejected, or whether it means it was accepted, or whether it means I have to do something else.

I may be able to find out myself, but do let me know if I should worry.

**Comment by kesterlester** [comment]
I am reopening as I think I may have accidentally closed it! Not sure how.

File: science/hacktv/Portfile
Comment (catap) [reply science/hacktv/Portfile:30]: Not worry, you're quite close :)

The only things that I see is missed is `CFLAGS`, and I suggest to patch Makefile to enjoy a PortGroup which takes care of it.

File: science/hacktv/files/destdir-fix.diff
Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff]: changed

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile]: I think I did it. But I really have to do something else now. Working week ahead.

File: science/hacktv/Portfile
Comment (catap) [inline science/hacktv/Portfile:33]: Please, remove useless comments

File: science/hacktv/Portfile
Comment (catap) [inline science/hacktv/Portfile:29]: May you move PortGroup to begining of the file? To like 6 for example.

File: science/hacktv/files/destdir-fix.diff
Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff]: ok

File: science/hacktv/files/destdir-fix.diff
Comment (reneeotten) [inline science/hacktv/files/destdir-fix.diff:29]: this path should not be hardcoded as users are free to install MacPorts in another prefix

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile:33]: Really?  I put them there as I thought I might have to maintain this in future, and 6 months from now I'll never remember what defines CC unless I have a comment like that in there.  I mean, if someone else maintains it and does loads of Portfile maintaining they wouldn't need that comment. But realistically I will need it -- or rather it will drastically reduce the time I spend having to work out what to do.

File: science/hacktv/Portfile
Comment (kesterlester) [reply science/hacktv/Portfile:29]: Again, I can see that PortGroups are  often at the top. But I wanted the big comment near the PortGroup command (the one you objected to above) and it looked way too ugly at the top of the file, so I put both down here.

File: science/hacktv/Portfile
Comment (catap) [reply science/hacktv/Portfile:33]: @reneeotten?

File: science/hacktv/Portfile
Comment (catap) [reply science/hacktv/Portfile:29]: @reneeotten?

File: science/hacktv/files/destdir-fix.diff
Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:29]: OK - help me here. I must have misunderstood the guidance I read doing that. I thought that DESTDIR was the very variable which (directly or indirectly) would be being controlled by a user trying to use other prefixes.  Is that not the case?  More concretely, perhaps I should ask just what should it look like? Is it like :

`install -c hacktv $(DESTDIR)/$(PREFIX)/bin'

? That would not have worked when I started this yesterday as (at that point) I didn't have a `PortGroup           makefile 1.0` line. However, I can imagine that maybe the existence of that line will mean that there are more env-vars in Make's envionment than I saw yesterday when PREFIX was not defined.

Comment (catap) [reply science/hacktv/files/destdir-fix.diff:29]: @kesterlester personally I've simple use `destroot` block inside portfile to copy an artifact to a target.

Your way requires modify a Makefile from Portfile to adjust to real `${prefix}`.

Comment (kesterlester) [reply science/hacktv/files/destdir-fix.diff:29]: ok - thank you for the tip. I may try this later in the week.

File: science/hacktv/Portfile
Comment (reneeotten) [reply science/hacktv/Portfile:29]: please follow the conventions: PortGroups at the top

File: science/hacktv/Portfile
Comment (reneeotten) [reply science/hacktv/Portfile:33]: I don't see a point for these comments either. The `makefile` PortGroup defines these things so you will not need to remember, that's the whole point.

File: science/hacktv/files/destdir-fix.diff
Comment (catap) [reply science/hacktv/files/destdir-fix.diff:29]: @kesterlester other way, probably cleaner, is using `DESTDIR` here.

**Comment by kesterlester** [comment]
Am closing this request without merging (i.e. am withdrawing my merge request) as I am no longer convinced it's really worth the effort for me to carry on.

In order to get this far  I have (naturally) had to create a local portfile repository in order to test that what I've been submitting here works. That local repository really creates for me 90% of the future benefits I'd have got from this merge request being accepted -- so the only thing I would be doing here by continuing this merge would be altruistic.  I was very happy to continue that for some time (I have been a beneficiary of others' macports altruism in the past, so it's only reasonable for me to give back).  

However:  now that I can see that I won't be allowed to put comments in a Portfile that I would like to be present to assist me in providing any  future maintenance, I am no longer convinced that this is the place for me. Knowing myself, and knowing my own limitations, I know I need to often put comments in code that I will make use of in six months time as pointers.  My understanding of this review process is that I will not be able to put such comments in a Portfile. The consequence of this is that I am tipped over the edge of no longer being willing to be altruistic.

I would have carried on if I felt I could document my own work to the extent I would want it documented to assist my future contributions ..... but I am not willing to carry on within a process that commits me to making my own future maintenance contributions harder than they could easily be.

**Pull request closed without merging**
